Feat: implement TTL for both SETs

Emil-Stampfly-He opened the pull request:
Hi Lorenzo, I'm not sure if this repo is still in maintenance but anyways this PR closes Issue #6 :)

## Update
1. Fixed typo \
`socekt` -> `socket` in `server.rs`; \
"adapatively" -> "adaptively" in README.md;
2. Updated README.md for the new feature
3. Implemented TTL for `Set` and `MultipleSet`

## Potential Enhancement
Currently if using `MultipleSet` then every entry will be set with the same TTL. For future enhancement TTL could be varied in `MultipleSet`.

Emil-Stampfly-He committed:
Successfully add TTL for both SETs

lorenzoc25 commented on readme.md:
```diff
@@ -26,26 +26,31 @@ And the server will be listening on the port you specified or `127.0.0.1:6379` b
+
+# invalid because ttl have to be specified
+$ curl 'localhost:6379/set/hello/world'
+{"SET": "Invalid"}
```
Let's provide a default value for argument here.

lorenzoc25 commented on src/command/mod.rs:
```diff
@@ -152,6 +160,14 @@ fn split_on_path(input: &str) -> Vec<&str> {
+        Ok(v) => Ok(v),
+        Err(e) => {
+            eprintln!("JSON parse error: {}", e);
+            eprintln!("bytes dump: {:?}", bytes);
```
not sure if we want to leave the bytes dump here. are you using it for your local debugging purpose?

lorenzoc25 commented on src/server.rs:
```diff
@@ -6,9 +6,14 @@ use std::collections::HashMap;
+    // key: (key, ttl) value: created time
+    Arc::new(Mutex::new(HashMap::new()))
+});
+
```
ugh oh, this might lead to an issue. 

you're using `(String, u64)` as the key in EXPIRE_MAP. But a key should only have one TTL at a time.

This approach allows multiple `(key, ttl)` entries to accumulate if the same key is set with different TTLs over time.

A better idea would be just map string to (start_time, ttl)

lorenzoc25 commented on src/server.rs:
```diff
@@ -69,7 +74,21 @@ async fn process(socket: TcpStream, db: ShardedDb) {
+
+                    // delete from EXPIRE_MAP & db if expired
+                    let now = current_unix_timestamp();
+                    let mut expire_map = EXPIRE_MAP.lock().unwrap();
```
Both `EXPIRE_MAP` and each shard's db are locked in a nested way during every Get here.
Under high concurrency, this might be problematic.

lorenzoc25 reviewed (changes requested):
Thanks for the contribution — this is a solid starting point! I have a few suggestions that could help improve it:

1. It looks like there isn’t a unit test for this logic yet. Could you add one to make sure the TTL behavior works as expected?
2. Inside the implementation, could we explicitly clarify the unit of TTL? Right now it’s using milliseconds, but that’s only mentioned in the doc. Making it clear in the code would help avoid confusion.
3. It might be worth considering a background cleanup task. I know that this is a mini project, but in real-world scenarios, Redis-like stores tend to be more read-heavy than write-heavy. Doing cleanup during reads might introduce performance or correctness issues.

Appreciate your work on this — thanks again :)

Emil-Stampfly-He commented:
> Thanks for the contribution — this is a solid starting point! I have a few suggestions that could help improve it:
> 
> 1. It looks like there isn’t a unit test for this logic yet. Could you add one to make sure the TTL behavior works as expected?
> 2. Inside the implementation, could we explicitly clarify the unit of TTL? Right now it’s using milliseconds, but that’s only mentioned in the doc. Making it clear in the code would help avoid confusion.
> 3. It might be worth considering a background cleanup task. I know that this is a mini project, but in real-world scenarios, Redis-like stores tend to be more read-heavy than write-heavy. Doing cleanup during reads might introduce performance or correctness issues.
> 
> Appreciate your work on this — thanks again :)

Thanks for your prompt and brilliant suggestions and review! Will be looking into them deeply asap.

As for the deletion strategy I originally planned to implement lazy deletion at first, and then maybe add active deletion for the future enhancement.

Emil-Stampfly-He committed:
Fixed expiration-cleanup logic and lock ordering

- Add active deletion of expired keys
- Enforce consistent lock order to prevent deadlocks

Emil-Stampfly-He committed:
Updated readme.md

Emil-Stampfly-He commented:
Hi again, I've fixed a few problems and added something new to it.

1. Deleted `bytes dump` & `Body as text` \
Yes, both of them are for local debugging usage.
2. Changed `EXPIRE_MAP` \
`(key, ttl) -> created_time` to `key -> (created_time, ttl)`
3. Changed name `ttl` to `ttl_ms` everywhere \
I'm not so sure about how it means "explicitly clarify the unit of TTL" that you mentioned here. Another way might be to use `Duration::from_millis` but name changing is the easiest way.
4. Fixed the problems with nested lock 
    - All operations now always lock `EXPIRE_MAP` and release it before locking the appropriate db shard.
    - On `GET`: check and remove expired entries under the `EXPIRE_MAP` lock, then explicitly drop that lock before acquiring the db shard lock to delete the data—never holding both at once.
5. Updated `readme.md` file
6. Added unit tests
7. Added active deletion strategy
8. Added default TTL \
`u64::MAX`, which means it never expires if TTL is not specified.

jayvdb commented:
ping @lorenzoc25
